gato/policy/gato_policy.py

```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
+
+        response_tokens = []
+
+        for idx in range(max_length): # idx can be viewed as the index of the 'next_token' within the list of such generated response tokens
+
+            # Include 'image_embeddings' instead of 'images' below to avoid re-calculating the same image embedding in every loop
+            # Note: the following 'text' is actually text token IDs instead of actual text. 
+            # We were using actual text before (and at that time tokenize_input_dicts(...) function also took actual text), and found an error 
+            # due to an edge case. The edge case is that the generated tokens could contain a sequence of .... (periods) or ,,,, (commas), 
+            # and if actual text is used in the following 'text', the tokenizer will treat the sequence as one token (which is not the desired behavior) 
+            # instead of a group of separate tokens (the desired behavior), and raise "index out of bound" exception on "n_patches - 1 + len(prompt_tokens)" 
+            # below where next_token_logits is calculated
+            batch_dict = {
+                'image_embeddings': image_embeddings,
+                'text': torch.tensor(prompt_tokens + response_tokens)
+            }
+
+            logits, _ = self.forward([batch_dict])
+            # logits' shape is (batch_size, seq_length, vocab_size), here batch_size should always be 1
+            assert logits.shape[0] == 1, "batch size should always be 1 for predicting response"
+
+            # Each patch of an image is treated as one token (vision transformer), so n_patches is the number of tokens representing an image
+            # In the model training (fine-tuning) for image-response task, each sequence starts with image tokens, followed by the text tokens 
+            # of the corresponding response. When predicting the response from an image, we pass the image embeddings of the image tokens through 
+            # the transformer, and then auto-regressively predict the next tokens, the context info of the image is conveyed through the 
+            # self-attention mechanism. The first text token of the generated response is the "next token" predicted by the last image token 
+            # in the sequence, and the second text token is predicted by the first text token, so on and so forth. 
+            # In the following line of code:
+            #   n_patches - 1 + len(prompt_tokens) + idx is the index of the token in the sequence that will predict the next text token of the generated response
+            #   squeeze the logits from shape (1, 1, vocab_size) to shape (vocab_size) and restrict the choice to the range 
+            #       within text tokens. The complete vocab contains non-text toksns also, we need to to exclude them from the choice
+            next_token_logits = logits[:, n_patches -1 + len(prompt_tokens) + idx, :].squeeze()[:self.text_tokens]     # shape (self.text_tokens)
```

Comment:
This is probably incorrect slicing. I will debug to check but it looks quite different from what's happening within text.